packages/client/src/lobby/hypoCompress.ts

```diff
@@ -0,0 +1,417 @@
+ *
+ * @param JSONString A string representing a hypothetical JSON.
+ */
+export function shrink(JSONString: string): string | null {
+  let gameDataJSON: GameJSON;
+  try {
+    gameDataJSON = JSON.parse(JSONString) as GameJSON;
+  } catch {
+    return null;
+  }
+
+  const compressed = gameJSONCompress(gameDataJSON);
+  if (compressed === null) {
+    return null;
+  }
+
+  // Check if compression succeeded
+  const decompressed = expand(compressed);
+  if (decompressed === null) {
+    return null;
+  }
+  try {
+    const decompressedJSON = JSON.parse(decompressed) as GameJSON;
+    if (!isEqual(gameDataJSON, decompressedJSON)) {
+      return null;
+    }
+  } catch {
+    return null;
+  }
+
+  return compressed;
+}
+
+/**
+ * Decompresses a string into a GameJSON object.
+ * Returns null if decompression fails.
+ *
+ * @param data The compressed string.
+ */
+export function expand(data: string): string | null {
+  // Remove all hyphens from URL
+  const normal = data.replace(/-/g, "");
+
+  // The compressed string is composed of 3 substrings separated by commas
+  const [playersAndDeck, actionsString, variantIDString] = [
+    ...normal.split(",", 3),
+  ];
+  const numberPlayersString = playersAndDeck.charAt(0);
+  const numPlayers = parseIntSafe(numberPlayersString);
+
+  const players = getPlayers(numPlayers);
+  if (players.length === 0) {
+    return null;
+  }
+
+  const deckString = playersAndDeck.substring(1);
+  const deck = decompressDeck(deckString);
+  if (deck === null) {
+    return null;
+  }
+
+  const actions = decompressActions(actionsString);
+
+  const variantID = parseIntSafe(variantIDString);
+  if (Number.isNaN(variantID)) {
+    return null;
+  }
+  const variant = getVariantByID(variantID);
+
+  const original: GameJSON = {
+    players,
+    deck,
+    actions,
+    options: {
+      variant: variant.name,
+    },
+    characters: [],
+    id: 0,
+    notes: [],
+    seed: "",
+  };
+
+  return JSON.stringify(original);
+}
+
+/**
+ * Compresses a GameJSON object into a string.
+ * Returns null if decompression fails.
+ *
+ * @param data The GameJSON object.
+ */
+function gameJSONCompress(data: GameJSON): string | null {
+  let out = "";
+
+  // Number of players
+  const numberOfPlayers = data.players.length;
+  if (numberOfPlayers < MIN_PLAYERS || numberOfPlayers > MAX_PLAYERS) {
+    return null;
+  }
+  out += `${data.players.length}`;
+
+  // Deck
+  const deck = compressDeck(data.deck);
+  if (deck === null) {
+    return null;
+  }
+  out += `${deck},`;
+
+  // Actions
+  const actions = compressActions(data.actions);
+  out += `${actions},`;
+
+  // Variant ID
+  if (data.options.variant === undefined) {
+    return null;
+  }
+
+  try {
+    const variant = getVariant(data.options.variant);
+    out += `${variant.id}`;
+  } catch {
+    return null;
+  }
+
+  // Add hyphens every 20 characters for URL posting (hyphens make the text wrap)
+  out = out.match(/.{1,20}/g)!.join("-");
```

Comment:
shouldn't we check for the error instead of using a null assertion operator?

Reply:
at that point, out is certainly non empty or the code would have returned null previously

Reply:
if you change any code, then out won't be guaranteed anymore to be 20 characters long, so it is fragile

Reply:
it doesn't have to. If out is less than 20 characters long, match will return an array with a single element (the entire out) and join will do nothing (effectively, with less than 20 characters no hyphen is added)

Reply:
then why is the null assertion operator necessary to have in there at all if the code will never be null

Reply:
because the compiler sees `match` and believes it might find nothing, which can never be the case here

Reply:
ok